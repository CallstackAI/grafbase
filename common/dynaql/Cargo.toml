[package]
name = "dynaql"
version = "3.0.31"
authors = ["sunli <scott_s829@163.com>", "Koxiaet"]
edition = "2021"
description = "A GraphQL server library implemented in Rust"
license = "MIT/Apache-2.0"
documentation = "https://docs.rs/async-graphql/"
homepage = "https://github.com/async-graphql/async-graphql"
repository = "https://github.com/async-graphql/async-graphql"
keywords = ["futures", "async", "graphql"]
categories = ["network-programming", "asynchronous"]
readme = "README.md"

[features]
default = ["query-planning"]
apollo_tracing = []
apollo_persisted_queries = ["lru", "sha2"]
unblock = ["blocking"]
string_number = []
dataloader = ["futures-timer", "futures-channel", "lru"]
tracing = ["tracinglib", "tracing-futures"]
decimal = ["rust_decimal"]
cbor = ["serde_cbor"]
chrono-duration = ["iso8601-duration"]
password-strength-validator = ["zxcvbn"]
unstable_oneof = ["dynaql-derive/unstable_oneof"]
tracing_worker = ["logworker"]
tokio_runtime = ["dynamodb/tokio_runtime"]
wasm_runtime = ["dynamodb/wasm_runtime"]
query-planning = ["dep:query-planning"]

[dependencies]
dynaql-derive = { path = "derive", version = "3" }
dynaql-value = { path = "value", version = "3" }
dynaql-parser = { path = "parser", version = "3" }

async-lock = "2.6"
async-recursion = "1"
async-stream = "0.3"
async-trait = "0.1"
bitflags = "1"
bytes = { version = "1", features = ["serde"] }
cached = { version = "0.42" }
chrono = { version = "0.4", default-features = false, features = [
  "clock",
  "std",
] }
derivative = "2"
dynomite = "0.10"
fast_chemail = "0.9"
fnv = "1"
futures-util = { version = "0.3", default-features = false, features = [
  "io",
  "sink",
] }
http = "0.2"
indexmap = { git = "https://github.com/bluss/indexmap.git", rev = "eedabaca9f84e520eab01325b305c08f3773e66c" }
ipnet = "2.7"
itertools = "0.10"
mime = "0.3"
multer = "2"
nom = "7.1"
num-traits = "0.2"
pin-project-lite = "0.2"
regex = "1"
send_wrapper = { version = "0.6", features = ["futures"] }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = { workspace = true }
static_assertions = "1"
tempfile = "3"
thiserror = { workspace = true }
ulid = { version = "1", features = ["wasm"] }
url = { version = "2" }
unindent = "0.2"
grafbase-runtime = { path = "../grafbase-runtime" }

# Query planning schema
arrow-schema = { version = "31", features = ["serde"] }

# API dependencies
query-planning = { path = "../query-planning", optional = true }
dynamodb = { path = "../dynamodb" }
graph-entities = { path = "../graph-entities" }
logworker = { path = "../log", package = "log", optional = true }

# Feature optional dependencies
bson = { version = "2", optional = true, features = ["chrono-0_4"] }
chrono-tz = { version = "0.8", optional = true }
hashbrown = { version = "0.13", optional = true }
iso8601-duration = { version = "0.1", optional = true }
log = { version = "0.4", optional = true }
secrecy = { version = "0.8", features = ["serde"] }
tracinglib = { version = "0.1", optional = true, package = "tracing" }
tracing-futures = { version = "0.2", optional = true, features = [
  "std-future",
  "futures-03",
] }
opentelemetry = { version = "0.17", optional = true, default-features = false, features = [
  "trace",
] }
uuid = { version = "1", optional = true, features = ["v4", "serde"] }
rust_decimal = { version = "1", optional = true }
smol_str = { version = "0.1", optional = true }
time = { version = "0.3", optional = true, features = [
  "parsing",
  "formatting",
  "macros",
] }
futures-timer = { version = "3", optional = true }
futures-channel = { version = "0.3", optional = true }

# Non-feature optional dependencies
blocking = { version = "1", optional = true }
lru = { version = "0.9", optional = true }
sha2 = { version = "0.10", optional = true }
serde_cbor = { workspace = true, optional = true }
zxcvbn = { version = "2", optional = true }

[target.'cfg(target_arch = "wasm32")'.dependencies]
surf = { version = "2", default-features = false, features = ["wasm-client"] }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
surf = { version = "2", default-features = false, features = [
    "h1-client-rustls",
] }

[dev-dependencies]
# tokio = { version = "1", features = ["macros", "rt-multi-thread", "sync", "time"] }
# futures-channel = "0.3"
insta = { version = "1", features = ["json"] }
once_cell = "1"
