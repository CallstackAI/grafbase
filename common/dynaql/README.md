# A GraphQL engine

A GraphQL engine written in Rust based on `async-graphql`.
