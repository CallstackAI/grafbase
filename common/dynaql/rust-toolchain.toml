[toolchain]
channel = "1.60"
targets = [
    "wasm32-unknown-unknown",     # Cloudflare Workers
]
components = ["rustfmt", "clippy"]
profile = "minimal"
