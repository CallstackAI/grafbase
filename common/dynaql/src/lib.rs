//! # A GraphQL server library implemented in Rust
#![deny(clippy::all)]
#![deny(clippy::inefficient_to_string)]
#![deny(clippy::match_wildcard_for_single_variants)]
#![deny(clippy::redundant_closure_for_method_calls)]
#![allow(clippy::cast_lossless)]
#![allow(clippy::cast_possible_truncation)]
#![allow(clippy::cast_sign_loss)]
#![allow(clippy::cognitive_complexity)]
#![allow(clippy::default_trait_access)]
#![allow(clippy::doc_markdown)]
#![allow(clippy::explicit_iter_loop)]
#![allow(clippy::future_not_send)]
#![allow(clippy::if_not_else)]
#![allow(clippy::implicit_hasher)]
#![allow(clippy::map_flatten)]
#![allow(clippy::map_unwrap_or)]
#![allow(clippy::manual_let_else)]
#![allow(clippy::match_same_arms)]
#![allow(clippy::missing_const_for_fn)]
#![allow(clippy::missing_errors_doc)]
#![allow(clippy::module_name_repetitions)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::needless_borrow)]
#![allow(clippy::needless_pass_by_value)]
#![allow(clippy::needless_lifetimes)]
#![allow(clippy::option_if_let_else)]
#![allow(clippy::panic)]
#![allow(clippy::redundant_else)]
#![allow(clippy::redundant_pub_crate)]
#![allow(clippy::return_self_not_must_use)]
#![allow(clippy::semicolon_if_nothing_returned)]
#![allow(clippy::similar_names)]
#![allow(clippy::too_many_lines)]
#![allow(clippy::trivially_copy_pass_by_ref)]
#![allow(clippy::unused_self)]
#![allow(clippy::upper_case_acronyms)]
#![allow(clippy::use_self)]
#![allow(clippy::useless_let_if_seq)]
#![allow(elided_lifetimes_in_paths)]
#![recursion_limit = "256"]
#![forbid(unsafe_code)]
#![cfg_attr(docsrs, feature(doc_cfg))]

mod base;
mod custom_directive;
mod error;
mod guard;
mod look_ahead;
#[doc(hidden)]
pub mod model;
pub mod names;
mod request;
mod response;
mod schema;
mod subscription;
pub mod validation;

pub mod auth;
pub mod context;
pub mod extensions;
pub mod http;
#[cfg(feature = "query-planning")]
pub mod logical_plan_utils;

pub mod resolver_utils;
pub mod types;
#[doc(hidden)]
pub mod validators;

pub mod graph;

pub mod registry;

#[doc(hidden)]
pub use async_stream;
#[doc(hidden)]
pub use async_trait;
#[doc(hidden)]
pub use context::ContextSelectionSet;
#[doc(hidden)]
pub use futures_util;
#[doc(hidden)]
pub use graph_entities;
#[doc(hidden)]
pub use indexmap;
#[doc(hidden)]
pub use static_assertions;
#[doc(hidden)]
pub use subscription::SubscriptionType;

pub use auth::*;
pub use base::{
    ComplexObject, Description, InputObjectType, InputType, InterfaceType, ObjectType, OutputType,
    UnionType,
};
pub use custom_directive::{CustomDirective, CustomDirectiveFactory};
pub use dynaql_parser as parser;
pub use dynaql_value::{
    from_value, to_value, value, ConstValue as Value, DeserializerError, Name, Number,
    SerializerError, Variables,
};
pub use error::{
    Error, ErrorExtensionValues, ErrorExtensions, InputValueError, InputValueResult,
    ParseRequestError, PathSegment, Result, ResultExt, ServerError, ServerResult,
};
pub use extensions::ResolveFut;
pub use graph_entities::ResponseNodeId;
pub use guard::{Guard, GuardExt};
pub use look_ahead::Lookahead;
pub use registry::{CacheControl, CacheInvalidation};
pub use request::{BatchRequest, Request};
#[doc(no_inline)]
pub use resolver_utils::{ContainerType, EnumType, ScalarType};
pub use response::GraphQlResponse;
pub use response::{BatchResponse, Response};
pub use schema::{Schema, SchemaBuilder, SchemaEnv};
pub use validation::{ValidationMode, ValidationResult, VisitorContext};
pub use validators::CustomValidator;

pub use context::*;
#[doc(no_inline)]
pub use parser::{Pos, Positioned};
pub use types::*;

/// An alias of [dynaql::Error](struct.Error.html). Present for backward compatibility
/// reasons.
pub type FieldError = Error;

/// An alias of [dynaql::Result](type.Result.html). Present for backward compatibility
/// reasons.
pub type FieldResult<T> = Result<T>;

#[doc = include_str!("docs/complex_object.md")]
pub use dynaql_derive::ComplexObject;
#[doc = include_str!("docs/description.md")]
pub use dynaql_derive::Description;
#[doc = include_str!("docs/directive.md")]
pub use dynaql_derive::Directive;
#[doc = include_str!("docs/enum.md")]
pub use dynaql_derive::Enum;
#[doc = include_str!("docs/input_object.md")]
pub use dynaql_derive::InputObject;
#[doc = include_str!("docs/interface.md")]
pub use dynaql_derive::Interface;
#[doc = include_str!("docs/merged_object.md")]
pub use dynaql_derive::MergedObject;
#[doc = include_str!("docs/merged_subscription.md")]
pub use dynaql_derive::MergedSubscription;
#[doc = include_str!("docs/newtype.md")]
pub use dynaql_derive::NewType;
#[doc = include_str!("docs/object.md")]
pub use dynaql_derive::Object;
#[cfg(feature = "unstable_oneof")]
#[cfg_attr(docsrs, doc(cfg(feature = "unstable_oneof")))]
#[doc = include_str!("docs/oneof_object.md")]
pub use dynaql_derive::OneofObject;
#[doc = include_str!("docs/scalar.md")]
pub use dynaql_derive::Scalar;
#[doc = include_str!("docs/simple_object.md")]
pub use dynaql_derive::SimpleObject;
#[doc = include_str!("docs/subscription.md")]
pub use dynaql_derive::Subscription;
#[doc = include_str!("docs/union.md")]
pub use dynaql_derive::Union;
