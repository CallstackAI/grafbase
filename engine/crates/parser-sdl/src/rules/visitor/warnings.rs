use std::{collections::BTreeSet, fmt};

use colored::Colorize;

#[derive(Clone, Debug, Default)]
pub(crate) struct Warnings {
    warnings: BTreeSet<Warning>,
}

impl Warnings {
    #[allow(dead_code)]
    pub fn push(&mut self, warning: Warning) {
        self.warnings.insert(warning);
    }

    pub fn is_empty(&self) -> bool {
        self.warnings.is_empty()
    }
}

#[derive(Clone, Debug, Ord, PartialOrd, PartialEq, Eq)]
pub(crate) enum Warning {}

impl AsRef<str> for Warning {
    fn as_ref(&self) -> &str {
        unreachable!()
    }
}

impl fmt::Display for Warnings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}", "Warnings:".bold().yellow())?;

        for warning in &self.warnings {
            writeln!(f, "  - {}", warning.as_ref().yellow())?;
        }

        Ok(())
    }
}
