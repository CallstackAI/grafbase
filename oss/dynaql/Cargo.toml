[package]
name = "dynaql"
version = "3.0.31"
edition = "2021"
description = "A dynamic GraphQL engine written in Rust"
repository = "https://github.com/grafbase/api"
readme = "README.md"

[features]
default = []
tracing_worker = ["logworker", "dynamodb/tracing", "tracing"]

[dependencies]
async-lock = "2"
async-recursion = "1"
async-stream = "0.3"
async-trait = "0.1"
bytes = { version = "1", features = ["serde"] }
case = "1"
chrono = { version = "0.4", default-features = false, features = [
  "clock",
  "std",
] }
cynic-introspection = "3"
derivative = "2"
dynomite = "0.10"
fast_chemail = "0.9"
flexbuffers = { workspace = true }
fnv = "1"
futures-util = { workspace = true, features = ["io", "sink"] }
http = "0.2"
indexmap = { git = "https://github.com/bluss/indexmap.git", rev = "eedabaca9f84e520eab01325b305c08f3773e66c" }
indoc = "2"
Inflector = { version = "0.11" }
internment = { version = "0.7", features = ["serde", "arc"] }
ipnet = "2"
itertools = { workspace = true }
mime = "0.3"
multer = "2"
nom = "7"
num-traits = "0.2"
once_cell = "1"
pin-project-lite = "0.2"
regex = { workspace = true }
secrecy = { version = "0.8", features = ["serde"] }
send_wrapper = { version = "0.6", features = ["futures"] }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = { workspace = true }
static_assertions = "1"
tempfile = "3"
thiserror = { workspace = true }
ulid = { workspace = true, features = ["wasm"] }
url = { workspace = true, features = ["serde"] }

# API dependencies
dataloader = { workspace = true }
dynamodb = { workspace = true }
dynaql-derive = { path = "derive", version = "3" }
dynaql-value = { path = "value", version = "3" }
dynaql-parser = { path = "parser", version = "3" }
grafbase-runtime = { workspace = true }
grafbase = { path = "../grafbase", features = ["json"] }
graph-entities = { workspace = true }
logworker = { path = "../log", package = "log", features = [
  "with-worker",
], optional = true }

# Feature optional dependencies
rust_decimal = { version = "1" }
tracing = { workspace = true, optional = true }
async-runtime = { workspace = true }

# Non-feature optional dependencies
base64.workspace = true

[target.'cfg(target_arch = "wasm32")'.dependencies]
reqwest = { version = "0.11", default-features = false, features = ["json"] }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
reqwest = { version = "0.11", default-features = false, features = [
  "json",
  "rustls-tls",
] }

[dev-dependencies]
indoc = "2"
insta = { version = "1", features = ["json"] }
rstest = { workspace = true }
sanitize-filename = "0.5"
tokio = { workspace = true, features = ["macros"] }
urlencoding = "2.1"
wiremock = "0.5"
