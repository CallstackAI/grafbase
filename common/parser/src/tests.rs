use dynaql::Schema;
use serde_json as _;

use crate::rules::visitor::RuleError;

macro_rules! assert_validation_error {
    ($schema:literal, $expected_message:literal) => {
        assert_matches!(
            super::to_registry($schema)
                .err()
                .and_then(super::Error::validation_errors)
                // We don't care whether there are more errors or not.
                // It only matters that we find the expected error.
                .and_then(|errors| errors.into_iter().next()),
            Some(RuleError { message, .. }) if message == $expected_message
        );
    };
}

#[test]
fn test_simple_product() {
    let result = super::to_registry(
        r#"
        type Product @model {
            id: ID!
            name: String!
            """
            The product's price in $
            """
            price: Int!
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_simple_todo() {
    let result = super::to_registry(
        r#"
        type Todo @model {
          id: ID!
          content: String!
          author: Author
        }

        type Author {
          name: String!
          lastname: String!
          pseudo: String
          truc: Truc!
        }

        type Truc {
          name: String!
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_simple_todo_from_template() {
    let result = super::to_registry(
        r#"
        type TodoList @model {
          id: ID!
          title: String!
          todos: [Todo]
        }

        type Todo @model {
          id: ID!
          title: String!
          complete: Boolean
        }
        "#,
    )
    .unwrap();

    let sdl = Schema::new(result).sdl();

    insta::assert_snapshot!(sdl);
}

#[test]
fn test_simple_todo_with_vec() {
    let result = super::to_registry(
        r#"
        type Todo @model {
          id: ID!
          content: String!
          authors: [Author]
        }

        type Author {
          name: String!
          lastname: String!
          pseudo: String
          truc: Truc!
        }

        type Truc {
          name: String!
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_simple_todo_with_enum() {
    let result = super::to_registry(
        r#"
        """
        A TodoType
        """
        enum TodoType {
          TODO1

          """
          A Type 2 for TODO
          """
          TODO2
        }

        type Todo @model {
          id: ID!
          content: String!
          authors: [Author]
          ty: TodoType!
        }

        type Author {
          name: String!
          lastname: String!
          pseudo: String
          truc: Truc!
        }

        type Truc {
          name: String!
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_simple_post_with_relation() {
    let result = super::to_registry(
        r#"
        enum Country {
          FRANCE
          NOT_FRANCE
        }

        type Blog @model {
          id: ID!
          posts: [Post] 
          owner: Author!
        }

        type Post @model {
          id: ID!
          content: String!
          authors: [Author] @relation(name: "published")
        }

        type Author @model {
          id: ID!
          name: String!
          lastname: String!
          country: Country!
          posts: [Post] @relation(name: "published")
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_multiple_relations() {
    let result = super::to_registry(
        r#"
        type Author @model {
          id: ID!
          lastname: String!
          published: [Post] @relation(name: "published")
          commented: [Comment] @relation(name: "commented")
        }

        type Post @model {
          id: ID!
          content: String!
          author: Author @relation(name: "published")
          comments: [Comment] @relation(name: "comments")
        }

        type Comment @model {
          id: ID!
          author: Author! @relation(name: "commented")
          post: Post @relation(name: "comments")
          comment: String!
          like: Int!
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn test_many_to_many() {
    let result = super::to_registry(
        r#"
        type User @model {
            name: String!
            organizations: [Organization!]
        }

        type Organization @model {
          name: String!
          users: [User!]
        }
        "#,
    )
    .unwrap();

    let reg_string = serde_json::to_value(&result).unwrap();
    let sdl = Schema::new(result).sdl();

    insta::assert_json_snapshot!(reg_string);
    insta::assert_snapshot!(sdl);
}

#[test]
fn should_ensure_lowercase() {
    let result = super::to_registry(
        r#"
        type Blog @model {
          id: ID!
          truc_break: String! @unique
        }
        "#,
    );

    assert!(result.is_err(), "Should error here");
}

#[test]
fn test_model_reserved_fields() {
    let with_reserved_fields = super::to_registry(
        r#"
        type Product @model {
            title: String
            id: ID!
            createdAt: DateTime!
            updatedAt: DateTime!
        }
        "#,
    )
    .unwrap();

    let without_reserved_fields = super::to_registry(
        r#"
        type Product @model {
            title: String
        }
        "#,
    )
    .unwrap();

    let req_string_a = serde_json::to_value(&with_reserved_fields).unwrap();
    let sdl_a = Schema::new(with_reserved_fields).sdl();

    let req_string_b = serde_json::to_value(&without_reserved_fields).unwrap();
    let sdl_b = Schema::new(without_reserved_fields).sdl();

    // Comparing snaphots for better test errors first
    insta::assert_json_snapshot!(req_string_a);
    insta::assert_snapshot!(sdl_a);

    insta::assert_json_snapshot!(req_string_b);
    insta::assert_snapshot!(sdl_b);

    // Actual test, ensuring they're equivalent.
    assert_eq!(req_string_a, req_string_b);
    assert_eq!(sdl_a, sdl_b);
}

#[test]
fn should_ensure_reserved_fields_have_correct_type_if_present() {
    assert_validation_error!(
        r#"
        type Product @model {
            id: Int!
        }
        "#,
        "Field 'id' of 'Product' is reserved by @model directive. It must have the type 'ID!' if present."
    );

    assert_validation_error!(
        r#"
        type Dummy @model {
            createdAt: DateTime
        }
        "#,
        "Field 'createdAt' of 'Dummy' is reserved by @model directive. It must have the type 'DateTime!' if present."
    );

    assert_validation_error!(
        r#"
        type Product @model {
            updatedAt: String!
        }
        "#,
        "Field 'updatedAt' of 'Product' is reserved by @model directive. It must have the type 'DateTime!' if present."
    );

    assert!(
        super::to_registry(
            r#"
            type Product @model {
                id: ID!
                createdAt: DateTime!
                updatedAt: DateTime!
            }
            "#
        )
        .is_ok(),
        "Should support specifying explicitly reserved fields."
    );

    assert!(
        super::to_registry(
            r#"
            type Product @model {
                name: String
            }
            "#
        )
        .is_ok(),
        "@model directive should not require reserved fields."
    );

    assert!(
        super::to_registry(
            r#"
            type Product {
                id: Int
                createdAt: DateTime
                updatedAt: String!
            }
            "#
        )
        .is_ok(),
        "Reserved fields only apply with the @model directive."
    );
}

#[test]
fn should_have_unique_fields() {
    assert_validation_error!(
        r#"
        type Product {
            count: Int
            count: Int
        }
        "#,
        "Field 'count' cannot be defined multiple times."
    );
}
