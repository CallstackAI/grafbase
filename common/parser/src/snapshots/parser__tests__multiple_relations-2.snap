---
source: common/parser/src/tests.rs
expression: sdl
---
type Author {
	lastname: String!
	published(after: String, before: String, first: Int, last: Int): PostConnection
	commented(after: String, before: String, first: Int, last: Int): CommentConnection
	"""
	Unique identifier
	"""
	id: ID!
	"""
	when the model was updated
	"""
	updatedAt: DateTime!
	"""
	when the model was created
	"""
	createdAt: DateTime!
}
input AuthorByInput {
	id: ID
}
type AuthorConnection {
	"""
	Information to aid in pagination
	"""
	pageInfo: PageInfo!
	edges: [AuthorEdge]
}
"""
Input to create a Author
"""
input AuthorCreateInput {
	lastname: String!
	published: [PublishedCreatePostRelation]
	commented: [CommentedCreateCommentRelation]
}
type AuthorCreatePayload {
	author: Author
}
type AuthorDeletePayload {
	deletedId: ID!
}
type AuthorEdge {
	node: Author!
	cursor: String!
}
"""
Input to update a Author
"""
input AuthorUpdateInput {
	lastname: String
	published: [PublishedUpdatePostRelation]
	commented: [CommentedUpdateCommentRelation]
}
type AuthorUpdatePayload {
	author: Author
}
type Comment {
	author: Author!
	post: Post
	comment: String!
	like: Int!
	"""
	Unique identifier
	"""
	id: ID!
	"""
	when the model was updated
	"""
	updatedAt: DateTime!
	"""
	when the model was created
	"""
	createdAt: DateTime!
}
input CommentByInput {
	id: ID
}
type CommentConnection {
	"""
	Information to aid in pagination
	"""
	pageInfo: PageInfo!
	edges: [CommentEdge]
}
"""
Input to create a Comment
"""
input CommentCreateInput {
	author: CommentedCreateAuthorRelation!
	post: CommentsCreatePostRelation
	comment: String!
	like: Int!
}
type CommentCreatePayload {
	comment: Comment
}
type CommentDeletePayload {
	deletedId: ID!
}
type CommentEdge {
	node: Comment!
	cursor: String!
}
"""
Input to update a Comment
"""
input CommentUpdateInput {
	author: CommentedUpdateAuthorRelation
	post: CommentsUpdatePostRelation
	comment: String
	like: IntOperationsInput
}
type CommentUpdatePayload {
	comment: Comment
}
"""
Input to create a Author for the commented relation of Comment
"""
input CommentedCreateAuthor {
	lastname: String!
	published: [PublishedCreatePostRelation]
	commented: [CommentedCreateCommentRelation]
}
"""
Input to link to or create a Author for the commented relation of Comment
"""
input CommentedCreateAuthorRelation {
	create: CommentedCreateAuthor
	link: ID
}
"""
Input to create a Comment for the commented relation of Author
"""
input CommentedCreateComment {
	post: CommentsCreatePostRelation
	comment: String!
	like: Int!
}
"""
Input to link to or create a Comment for the commented relation of Author
"""
input CommentedCreateCommentRelation {
	create: CommentedCreateComment
	link: ID
}
"""
Input to link/unlink to or create a Author for the commented relation of Comment
"""
input CommentedUpdateAuthorRelation {
	create: CommentedCreateAuthor
	link: ID
	unlink: ID
}
"""
Input to link/unlink to or create a Comment for the commented relation of Author
"""
input CommentedUpdateCommentRelation {
	create: CommentedCreateComment
	link: ID
	unlink: ID
}
"""
Input to create a Comment for the comments relation of Post
"""
input CommentsCreateComment {
	author: CommentedCreateAuthorRelation!
	comment: String!
	like: Int!
}
"""
Input to link to or create a Comment for the comments relation of Post
"""
input CommentsCreateCommentRelation {
	create: CommentsCreateComment
	link: ID
}
"""
Input to create a Post for the comments relation of Comment
"""
input CommentsCreatePost {
	content: String!
	author: PublishedCreateAuthorRelation
	comments: [CommentsCreateCommentRelation]
}
"""
Input to link to or create a Post for the comments relation of Comment
"""
input CommentsCreatePostRelation {
	create: CommentsCreatePost
	link: ID
}
"""
Input to link/unlink to or create a Comment for the comments relation of Post
"""
input CommentsUpdateCommentRelation {
	create: CommentsCreateComment
	link: ID
	unlink: ID
}
"""
Input to link/unlink to or create a Post for the comments relation of Comment
"""
input CommentsUpdatePostRelation {
	create: CommentsCreatePost
	link: ID
	unlink: ID
}
"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, is compliant with the date-time format outlined in section 5.6 of the RFC 3339
profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.

This scalar is a description of an exact instant on the timeline such as the instant that a user account was created.

# Input Coercion

When expected as an input type, only RFC 3339 compliant date-time strings are accepted. All other input values raise a query error indicating an incorrect type.

# Result Coercion

Where an RFC 3339 compliant date-time string has a time-zone other than UTC, it is shifted to UTC.
For example, the date-time string 2016-01-01T14:10:20+01:00 is shifted to 2016-01-01T13:10:20Z.
"""
scalar DateTime
"""
Possible operations for an Int field
"""
input IntOperationsInput {
	set: Int
	increment: Int
	decrement: Int
}
type Mutation {
	"""
	Create a Author
	"""
	authorCreate(input: AuthorCreateInput!): AuthorCreatePayload
	"""
	Update a Author
	"""
	authorUpdate(by: AuthorByInput!, input: AuthorUpdateInput!): AuthorUpdatePayload
	"""
	Delete a Author by ID or unique field
	"""
	authorDelete(by: AuthorByInput!): AuthorDeletePayload
	"""
	Create a Post
	"""
	postCreate(input: PostCreateInput!): PostCreatePayload
	"""
	Update a Post
	"""
	postUpdate(by: PostByInput!, input: PostUpdateInput!): PostUpdatePayload
	"""
	Delete a Post by ID or unique field
	"""
	postDelete(by: PostByInput!): PostDeletePayload
	"""
	Create a Comment
	"""
	commentCreate(input: CommentCreateInput!): CommentCreatePayload
	"""
	Update a Comment
	"""
	commentUpdate(by: CommentByInput!, input: CommentUpdateInput!): CommentUpdatePayload
	"""
	Delete a Comment by ID or unique field
	"""
	commentDelete(by: CommentByInput!): CommentDeletePayload
}
type PageInfo {
	hasPreviousPage: Boolean!
	hasNextPage: Boolean!
	startCursor: String
	endCursor: String
}
type Post {
	content: String!
	author: Author
	comments(after: String, before: String, first: Int, last: Int): CommentConnection
	"""
	Unique identifier
	"""
	id: ID!
	"""
	when the model was updated
	"""
	updatedAt: DateTime!
	"""
	when the model was created
	"""
	createdAt: DateTime!
}
input PostByInput {
	id: ID
}
type PostConnection {
	"""
	Information to aid in pagination
	"""
	pageInfo: PageInfo!
	edges: [PostEdge]
}
"""
Input to create a Post
"""
input PostCreateInput {
	content: String!
	author: PublishedCreateAuthorRelation
	comments: [CommentsCreateCommentRelation]
}
type PostCreatePayload {
	post: Post
}
type PostDeletePayload {
	deletedId: ID!
}
type PostEdge {
	node: Post!
	cursor: String!
}
"""
Input to update a Post
"""
input PostUpdateInput {
	content: String
	author: PublishedUpdateAuthorRelation
	comments: [CommentsUpdateCommentRelation]
}
type PostUpdatePayload {
	post: Post
}
"""
Input to create a Author for the published relation of Post
"""
input PublishedCreateAuthor {
	lastname: String!
	published: [PublishedCreatePostRelation]
	commented: [CommentedCreateCommentRelation]
}
"""
Input to link to or create a Author for the published relation of Post
"""
input PublishedCreateAuthorRelation {
	create: PublishedCreateAuthor
	link: ID
}
"""
Input to create a Post for the published relation of Author
"""
input PublishedCreatePost {
	content: String!
	comments: [CommentsCreateCommentRelation]
}
"""
Input to link to or create a Post for the published relation of Author
"""
input PublishedCreatePostRelation {
	create: PublishedCreatePost
	link: ID
}
"""
Input to link/unlink to or create a Author for the published relation of Post
"""
input PublishedUpdateAuthorRelation {
	create: PublishedCreateAuthor
	link: ID
	unlink: ID
}
"""
Input to link/unlink to or create a Post for the published relation of Author
"""
input PublishedUpdatePostRelation {
	create: PublishedCreatePost
	link: ID
	unlink: ID
}
type Query {
	"""
	Query a single Author by an ID or a unique field
	"""
	author(by: AuthorByInput!): Author
	"""
	Paginated query to fetch the whole list of `Author`.
	"""
	authorCollection(after: String, before: String, first: Int, last: Int): AuthorConnection
	"""
	Query a single Post by an ID or a unique field
	"""
	post(by: PostByInput!): Post
	"""
	Paginated query to fetch the whole list of `Post`.
	"""
	postCollection(after: String, before: String, first: Int, last: Int): PostConnection
	"""
	Query a single Comment by an ID or a unique field
	"""
	comment(by: CommentByInput!): Comment
	"""
	Paginated query to fetch the whole list of `Comment`.
	"""
	commentCollection(after: String, before: String, first: Int, last: Int): CommentConnection
}
schema {
	query: Query
	mutation: Mutation
}

